core/src/main/java/org/infinispan/lifecycle/ModuleLifecycle.java

```diff
@@ -22,7 +25,17 @@
  * @since 4.0
  */
 public interface ModuleLifecycle {
-    default void addDynamicMetadata(ModuleRepository.Builder moduleBuilder, GlobalConfiguration globalConfiguration) {}
+
+    interface ModuleBuilder {
```

Comment:
@danberindei I think this will be of interest

Reply:
Yeah, I didn't mean to make it part of the API because I already have some changes in mind.
I think I can remove `addDynamicMetadata()` and implement `ModuleMetadataBuilder` manually to obtain the same result.

Reply:
I also realised well before this was merged that the changes I've done here still did not fully remove the leakage and had another commit to fix that on another branch and forgot to move it to this PR. But I resurrected it an here it comes https://github.com/infinispan/infinispan/pull/7373